hs/src/Reach/Connector/ETH_Solidity.hs

```diff
@@ -1387,9 +1396,37 @@ apiDef who ApiInfo {..} = do
               Nothing -> []
   return $ vsep $ mk who_s body : alias
 
+genApiJump :: SLPart -> M.Map Int ApiInfo -> App Doc
+genApiJump p ms = do
+  let who = pretty $ bunpack p
+  let ai = fromMaybe (impossible "genApiJump") $ headMay $ M.elems ms
+  (_, argDefns, _, args, _) <- apiArgs ai
+  let whichs = map ai_which $ M.elems ms
+  let chk_which w = solBinOp "==" "current_step" $ pretty w
+  let chk_st = concatWith (\ l r -> l <> " || " <> r) $ map chk_which whichs
+  let require = solApply "require" [chk_st] <> semi
+  let mk w = solWhen (chk_which w) thn
+        where
+          thn = "return " <> solApply ("this." <> inst) args <> semi <> hardline
+          inst = "_" <> who <> pretty w
+  let go = vsep $ map (mk . fst) $ M.toAscList ms
+  ret <- funRetSig $ ai_ret_ty ai
+  let body = vsep $ require : [go]
+  return $ solFunction who argDefns ret $ body
+
 apiDefs :: ApiInfos -> App Doc
-apiDefs defs =
-  vsep <$> (mapM (uncurry apiDef) $ M.toList defs)
+apiDefs defs = do
+  let defL = M.toList defs
+  defs' <- concat <$> mapM (\ (p, ms) -> do
+                              let qualify = M.size ms > 1
+                              ds <- mapM (\ (_, a) -> apiDef p qualify a) $ M.toAscList ms
+                              case qualify of
+                                True  -> do
+                                  d <- genApiJump p ms
+                                  return $ d : ds
+                                False -> return ds
+                            ) defL
```

Comment:
My taste is to define this as a `let go =` or do `defs' <- forM defL $ \(p, ms) -> do ....` and then `return $ vsep $ concat defs'`